tracing/tracing.go

```diff
@@ -0,0 +1,136 @@
+	writerExporter *stdouttrace.Exporter
+}
+
+var _ trace.SpanExporter = &fileExporter{}
+
+func newFileExporter(file string) (*fileExporter, error) {
+	f, err := os.OpenFile(file, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		return nil, fmt.Errorf("opening %s: %w", file, err)
+	}
+	stdoutExporter, err := stdouttrace.New(stdouttrace.WithWriter(f))
+	if err != nil {
+		return nil, err
+	}
+	return &fileExporter{
+		writerExporter: stdoutExporter,
+		file:           f,
+	}, nil
+}
+
+func (e *fileExporter) ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error {
+	return e.writerExporter.ExportSpans(ctx, spans)
+}
+
+func (e *fileExporter) Shutdown(ctx context.Context) error {
+	if err := e.writerExporter.Shutdown(ctx); err != nil {
+		return err
+	}
+	if err := e.file.Close(); err != nil {
+		return fmt.Errorf("closing trace file: %w", err)
+	}
+	return nil
+}
+
+// noopShutdownTracerProvider wraps a TracerProvider with a no-op Shutdown method.
+type noopShutdownTracerProvider struct {
+	tp traceapi.TracerProvider
+}
+
+func (n *noopShutdownTracerProvider) Shutdown(ctx context.Context) error {
+	return nil
+}
+func (n *noopShutdownTracerProvider) Tracer(instrumentationName string, opts ...traceapi.TracerOption) traceapi.Tracer {
+	return n.tp.Tracer(instrumentationName, opts...)
+}
+
+type ShutdownTracerProvider interface {
+	traceapi.TracerProvider
+	Shutdown(ctx context.Context) error
+}
+
+// NewTracerProvider creates and configures a TracerProvider.
+func NewTracerProvider(ctx context.Context) (ShutdownTracerProvider, error) {
+	if os.Getenv("IPFS_TRACING") == "" {
+		return &noopShutdownTracerProvider{tp: traceapi.NewNoopTracerProvider()}, nil
+	}
+
+	options := []trace.TracerProviderOption{}
+
+	traceRatio := 1.0
+	if envRatio := os.Getenv("IPFS_TRACING_RATIO"); envRatio != "" {
+		r, err := strconv.ParseFloat(envRatio, 64)
+		if err == nil {
+			traceRatio = r
+		}
+	}
+	options = append(options, trace.WithSampler(trace.ParentBased(trace.TraceIDRatioBased(traceRatio))))
+
+	r, err := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String("go-ipfs"),
+			semconv.ServiceVersionKey.String(version.CurrentVersionNumber),
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+	options = append(options, trace.WithResource(r))
+
+	for envVar, builder := range exporterBuilders {
+		if val := os.Getenv(envVar); val != "" {
+			exporter, err := builder(ctx, val)
+			if err != nil {
+				return nil, err
+			}
+			options = append(options, trace.WithBatcher(exporter))
+		}
+	}
+
+	return trace.NewTracerProvider(options...), nil
+}
+
+// Span starts a new span using the standard IPFS tracing conventions.
+func Span(ctx context.Context, componentName string, spanName string, opts ...traceapi.SpanStartOption) (context.Context, traceapi.Span) {
+	return otel.Tracer("go-ipfs").Start(ctx, fmt.Sprintf("%s.%s", componentName, spanName), opts...)
```

Comment:
I think it is intended that the default tracer is used since we aren't changing the format of the traces. We already record the name of the system and the software used to record the trace in the Resource section. See, for example, this section of one trace. The otel documentation is not super clear but I feel that the presence of `Version` and `SchemaURL` fields in the `InstrumentationLibrary` object imply that this is related to the type of trace that has been produced.

```json
    "Resource": [
        {
            "Key": "service.name",
            "Value": {
                "Type": "STRING",
                "Value": "go-ipfs"
            }
        },
        {
            "Key": "service.version",
            "Value": {
                "Type": "STRING",
                "Value": "0.13.0-dev"
            }
        },
        {
            "Key": "telemetry.sdk.language",
            "Value": {
                "Type": "STRING",
                "Value": "go"
            }
        },
        {
            "Key": "telemetry.sdk.name",
            "Value": {
                "Type": "STRING",
                "Value": "opentelemetry"
            }
        },
        {
            "Key": "telemetry.sdk.version",
            "Value": {
                "Type": "STRING",
                "Value": "1.6.1"
            }
        }
    ],
    "InstrumentationLibrary": {
        "Name": "go-ipfs",
        "Version": "",
        "SchemaURL": ""
    }
```

Reply:
Fair enough. I don't feel strongly about it. It's possible that the SchemaURL/Version fields are intended to describe the schema of the tags that the library is writing.

Reply:
Yeah I'm not sure what those are for. This is all clearly communicated as experimental so that we have some wiggle room to adjust these over time as we learn more.